spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -0,0 +1,582 @@
+        portIdentifer,
+        channelIdentifier
+    )
+    
+    // upgrade is complete
+    // set channel to OPEN and remove unnecessary state
+    currentChannel.state = OPEN
+    provableStore.set(channelPath(portIdentifier, channelIdentifier), currentChannel)
+    provableStore.delete(timeoutPath(portIdentifier, channelIdentifier))
+    privateStore.delete(restorePath(portIdentifier, channelIdentifier))
+}
+```
+
+
+### Cancel Upgrade Process
+
+During the upgrade handshake a chain may cancel the upgrade by writing an error receipt into the error path and restoring the original channel to `OPEN`. The counterparty must then restore its channel to `OPEN` as well. A relayer can facilitate this by calling `CancelChannelUpgrade`:
+
+```typescript
+function cancelChannelUpgrade(
+    portIdentifier: Identifier,
+    channelIdentifier: Identifier,
+    errorReceipt: []byte,
+    proofUpgradeError: CommitmentProof,
+    proofHeight: Height,
+) {
+    // current channel is in UPGRADE_INIT or UPGRADE_TRY
+    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+    abortTransactionUnless(channel.state == UPGRADE_INIT || channel.state == UPGRADE_TRY)
+
+    abortTransactionUnless(!isEmpty(errorReceipt))
+
+    abortTransactionUnless(verifyUpgradeError(currentChannel, proofHeight, proofUpgradeError, currentChannel.counterpartyChannelIdentifier, errorReceipt))
+
+    // cancel upgrade
+    // and restore original conneciton
+    // delete unnecessary state
+    originalChannel = privateStore.get(restorePath(portIdentifier, channelIdentifier))
+    provableStore.set(channelPath(portIdentifier, channelIdentifier), originalChannel)
+
+    // delete auxilliary upgrade state
+    provableStore.delete(timeoutPath(portIdentifier, channelIdentifier))
+    privateStore.delete(restorePath(portIdentifier, channelIdentifier))
+
+    // call modules onChanUpgradeRestore callback
+    module = lookupModule(portIdentifier)
+    // restore callback must not return error since counterparty successfully upgraded
+    module.onChanUpgradeRestore(
+        portIdentifer,
+        channelIdentifier
+    )
+}
+```
+
+### Timeout Upgrade Process
+
+It is possible for the channel upgrade process to stall indefinitely on UPGRADE_TRY if the UPGRADE_TRY transaction simply cannot pass on the counterparty; for example, the upgrade feature may not be enabled on the counterparty chain.
+
+In this case, we do not want the initializing chain to be stuck indefinitely in the `UPGRADE_INIT` step. Thus, the `UpgradeInit` message will contain a `TimeoutHeight` and `TimeoutTimestamp`. The counterparty chain is expected to reject `UpgradeTry` message if the specified timeout has already elapsed.
+
+A relayer must then submit an `UpgradeTimeout` message to the initializing chain which proves that the counterparty is still in its original state. If the proof succeeds, then the initializing chain shall also restore its original channel and cancel the upgrade.
```

Comment:
We should be consistent about the existing messages throughout the spec:
- `ChannelUpgradeTimeoutMsg`
- `ChannelUpgradeCancelMsg`

This should also be the case for the connection update, i.e., 
- `ConnUpgradeTimeoutMsg`
- `ConnUpgradeCancelMsg`